### PR TITLE
chore(common): maintenance on build scripts - cd

### LINKED ISSUE #11324: chore(common): use `cd "$THIS_SCRIPT_PATH"` in build-utils.sh
We should always start in the same path as build.sh, so we don't have to boilerplate this in every script.

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #11324.

* Always `cd "$THIS_SCRIPT_PATH"`
* Remove unnecessary `cd` from all build.sh
* Remove unnecessary `set -eu` from all build.sh (and `# set -x`)
* Replace old build-utils.sh incantation in a few build.sh scripts
* Builder scripts all now use resources/build/builder.inc.sh, and other scripts use build-utils.sh: if you don't call `builder_describe` and `builder_parse`, then you are not a builder script!

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/server/build.sh [hunk 1]
```diff
@@ -5,13 +5,10 @@
 THIS_SCRIPT="$(readlink -f "${BASH_SOURCE[0]}")"
 . "${THIS_SCRIPT%/*}/../../../resources/build/build-utils.sh"
 ## END STANDARD BUILD SCRIPT INCLUDE
-EX_USAGE=64
```
Comment (mcdurdin) [inline developer/src/server/build.sh:8]: Unused legacy cruft!

**Comment by mcdurdin** [comment]
Okay, one problem here is that build-utils.sh is used for both builder scripts and other utility scripts, some of which are supposed to _not_ change folder (e.g. mkver.inc.sh).

Two ways to fix:
* Have a separate include that is intended to be _just_ for builder scripts. This involves some shuffling:
  * Rename build-utils.sh to build-utils.inc.sh. (this is just tidy up)
  * Builder scripts should source a new resources/build/builder.inc.sh, and other scripts should source resources/build/build-utils.inc.sh.
  * resources/builder.inc.sh remains unchanged (perhaps rename to disambiguate with resources/build/builder.inc.sh)
* Or, add a `BUILDER_NO_CD` flag to scripts that care about initial folder
  * alt. could `pushd`/`popd` but that looks uglier.

**Comment by mcdurdin** [comment]
> Let me just put it in my words to see if it makes sense. If you add the line cd "$THIS_SCRIPT_PATH" to builder.inc.sh. builder.inc.sh is sourced in build-utils.inc.sh. So now all the subprojects that are sourcing build-utils.inc.sh will have the directory changed to the script path pf the calling script. This is not desirable eg mkver.inc.sh in which you want it to be the directory it is currently in at the time you source build-utils.inc.sh .

Yep.

> I would go with BUILDER_NO_CD flag. That is because adding another script just to not do a cd seems confusing.
> What about If $THIS_SCRIPT_PATH was called $EXECUTE_PATH then for those scripts like that do not want to change directory set the $EXECUTE_PATH to pwd directory.

I kinda started to go this way then realised I'd need to fixup all the random scripts that include build-utils.sh to do this, and it would still leave us in the spot where api contract for builder scripts is vague. So I ended up:

1. Creating resources/build/builder.inc.sh which imports build-utils.sh and also does the `cd`
2. Updating all builder-compliant scripts to use that one
3. Remaining scripts continue to use build-utils.sh
4. resources/builder.inc.sh may be renamed in future to avoid confusion

**Comment by rc-swag** [comment]
> 4\. resources/builder.inc.sh may be renamed in future to avoid confusion

Yes I think that needs to happen, something that helps reflect what it does the `resources/build/builder.inc.sh` doesn't. Actually it is really the other way around isn't?. the later is an extension

**Comment by mcdurdin** [comment]
> > 4. resources/builder.inc.sh may be renamed in future to avoid confusion
> 
> Yes I think that needs to happen, something that helps reflect what it does the `resources/build/builder.inc.sh` doesn't. Actually it is really the other way around isn't?. the later is an extension

I think resources/builder.inc.sh should become resources/build/builder.implementation.inc.sh or something like that (better not to have it in resources/ directly).

File: common/linux/build.sh [hunk 2]
```diff
@@ -23,6 +18,11 @@ builder_describe "Keyman common Linux modules" \
 
 builder_parse "$@"
 
+if [[ $BUILDER_OS != linux ]]; then
```
Comment (mcdurdin) [inline common/linux/build.sh:21]: A builder script must always call `builder_describe` and `builder_parse`, so do our platform checks after that. This allows the script to be called with `--help` even when on the wrong platform, so user can get info about it.

File: common/mac/build.sh [hunk 3]
```diff
@@ -23,6 +18,11 @@ builder_describe "Keyman common mac modules" \
 
 builder_parse "$@"
 
+if [[ $BUILDER_OS != mac ]]; then
+  builder_echo grey "Platform is not macOS; skipping common/mac"
+  exit 0
+fi
```
Comment (mcdurdin) [inline common/mac/build.sh:24]: A builder script must always call `builder_describe` and `builder_parse`, so do our platform checks after that. This allows the script to be called with `--help` even when on the wrong platform, so user can get info about it.

**Review by jahorton — COMMENTED** [review]
Noting some of the previous conversation and the documentation I saw in the file-changes, I think we have a few removals of change-directory that need to get restored.

File: common/linux/build.sh [hunk 4]
```diff
@@ -23,6 +18,11 @@ builder_describe "Keyman common Linux modules" \
...
+  builder_echo grey "Platform is not linux; skipping common/linux"
+  exit 0
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply common/linux/build.sh:23]: Aiming not to change behaviour for these scripts. This check came in #8692.

File: ios/samples/KMSample1/build.sh [hunk 5]
```diff
@@ -6,9 +6,6 @@ THIS_SCRIPT="$(readlink -f "${BASH_SOURCE[0]}")"
 . "${THIS_SCRIPT%/*}/../../../resources/build/build-utils.sh"
 ## END STANDARD BUILD SCRIPT INCLUDE
 
-# This script runs from its own folder
-cd "$THIS_SCRIPT_PATH"
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply ios/samples/KMSample1/build.sh:10]: Hmm, they should be builder scripts ... adding them to the list!

Will revert this.

File: mac/build.sh [hunk 6]
```diff
@@ -12,9 +12,6 @@ KEYMAN_MAC_BASE_PATH="$KEYMAN_ROOT/mac"
 . "$KEYMAN_ROOT/resources/build/build-help.inc.sh"
 . "$KEYMAN_ROOT/mac/mac-utils.inc.sh"
 
-# This script runs from its own folder
-cd "$(dirname "$THIS_SCRIPT")"
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply mac/build.sh:16]: Will revert this.

**Review by jahorton — APPROVED** [review]

**Review by rc-swag — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.28-alpha](https://keyman.com/downloads/releases/alpha/18.0.28)

**Pull request merged**